### PR TITLE
Introduce service config treatNilableAsOptional for query and header params

### PR DESCRIPTION
**Pull request by TharmiganK**
## Purpose

This adds support for query params without value to have a nil value and also introduces a service configuration `treatNilableAsOptional` to support ideal and business logic level mapping. (See the related issue)

This configuration is applied to query and header parameters.

The default behaviour is set to `treatNilableAsOptional = true` (Business logic level mapping). The behaviour can be set to ideal mapping by : 
```ballerina
@http:ServiceConfig {
      treatNilableAsOptional : false
}
```

> Related to [`Mapping HTTP headers to resource function arguments`](https://github.com/ballerina-platform/ballerina-standard-library/issues/1928)

## Examples

```ballerina
// Default setting : Business logic level mapping
service /defaultqueryparamservice on new http:Listener(8050) {
    // nilable query parameter foo will be treated as optional
    resource function get test1(string? foo) returns string {
        return foo ?: "empty";
    }
   // nillable header parameter will be treated as optional
   resource function get test2(@http:Header string? foo) returns string {
        return foo ?: "empty";
    }
}

// Ideal mapping
@http:ServiceConfig {
      treatNilableAsOptional : false
}
service /idealqueryparamservice on new http:Listener(8051) {
    // nilable query parameter foo will be treated as compulsory
    resource function get test1(string? foo) returns string {
        return foo ?: "empty";
    }
   // nillable header parameter will be treated as compulsory
   resource function get test2(@http:Header string? foo) returns string {
        return foo ?: "empty";
    }
}
```

| Query             | Mapping (Default) | Mapping (Ideal) |
| ------------- | ------------------ | ------------- |
| `foo=bar`      | bar                          | bar                  |
| `foo`              | nil                           | nil                    |
| No query       | nil                           | Error: no query param value found for 'foo'  |

| Header           | Mapping (Default) | Mapping (Ideal) |
| ------------- | ------------------ | ------------- |
| `foo : bar`     | bar                          | bar                  |
| `foo :`            | nil                           | nil                    |
| No header     | nil                           | Error: no header value found for 'foo'  |

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests

### CODE REVIEW

File: ballerina-tests/tests/service_dispatching_query_param_binding_ideal_mapping_test.bal [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    if response is http:Response {
+        test:assertEquals(response.statusCode, 400);
+        assertTextPayload(response.getTextPayload(), "no query param value found for 'foo'");
+    } else {
+        test:assertFail(msg = "Found unexpected output type: " + response.message());
+    }
+}
+
+@test:Config {}
+function testIdealQueryParamBindingWithEmptyQueryParam() {
```
Comment (chamil321) [inline ballerina-tests/tests/service_dispatching_query_param_binding_ideal_mapping_test.bal:76]:
```suggestion
function testIdealQueryParamBindingWithNoQueryParamValue() {
```

File: native/src/main/java/io/ballerina/stdlib/http/api/HttpResource.java [hunk 2]
```diff
@@ -203,6 +204,14 @@ public void setTransactionInfectable(boolean transactionInfectable) {
         this.transactionInfectable = transactionInfectable;
     }
 
+    public void setTreatNilableAsOptional(boolean treatNilableAsOptional) {
+        this.treatNilableAsOptional = treatNilableAsOptional;
+    }
+
+    public boolean getTreatNilableAsOptional() {
```
Comment (chamil321) [inline native/src/main/java/io/ballerina/stdlib/http/api/HttpResource.java:211]:
```suggestion
    public boolean isTreatNilableAsOptional() {
```

File: native/src/main/java/io/ballerina/stdlib/http/api/HttpService.java [hunk 3]
```diff
@@ -73,6 +74,7 @@
     private String chunkingConfig;
     private String mediaTypeSubtypePrefix;
     private String introspectionResourcePath;
+    private boolean treatNilableAsOptional = true;
```
Comment (chamil321) [inline native/src/main/java/io/ballerina/stdlib/http/api/HttpService.java:77]: Do we need treatNilableAsOptional in this class? Won't that be enough to cache in the resource?

Comment (TharmiganK) [reply native/src/main/java/io/ballerina/stdlib/http/api/HttpService.java:77]: We get the configuration in `buildHttpService` method and set the value for resources from `updateResourceTree` method. If we remove this from service, we have to pass it through `buildHttpService`->`processResources`->`updateResourceTree`. WDYT?

File: ballerina-tests/tests/service_dispatching_query_param_binding_ideal_mapping_test.bal (hunk 1 above)
Comment (TharmiganK) [reply ballerina-tests/tests/service_dispatching_query_param_binding_ideal_mapping_test.bal]: Resolved in [`ee94230`](https://github.com/ballerina-platform/module-ballerina-http/pull/653/commits/ee94230d58b6a47ddad67b84ac4bcab5227ee6eb)

File: native/src/main/java/io/ballerina/stdlib/http/api/HttpResource.java (hunk 2 above)
Comment (TharmiganK) [reply native/src/main/java/io/ballerina/stdlib/http/api/HttpResource.java]: Resolved in [`ee94230`](https://github.com/ballerina-platform/module-ballerina-http/pull/653/commits/ee94230d58b6a47ddad67b84ac4bcab5227ee6eb)

File: native/src/main/java/io/ballerina/stdlib/http/api/HttpService.java (hunk 3 above)
Comment (TharmiganK) [reply native/src/main/java/io/ballerina/stdlib/http/api/HttpService.java]: Resolved in [`a572748`](https://github.com/ballerina-platform/module-ballerina-http/pull/653/commits/a57274874a0c3a0bc6357de23c9053771bc96b6e)

File: ballerina-tests/tests/service_dispatching_header_param_binding_ideal_mapping_test.bal [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    resource function get test2(@http:Header string? foo, int bar) returns json {
+        json responseJson = { value1: foo ?: "empty", value2: bar};
+        return responseJson;
+    }
+}
+
+@test:Config {}
+function testIdealHeaderParamBindingWithHeaderValue() {
+    http:Response|error response = headerBindingIdealClient->get("/headerparamservice/test1?foo=WSO2&bar=56", {"foo":"Ballerina"});
```
Comment (ayeshLK) [inline ballerina-tests/tests/service_dispatching_header_param_binding_ideal_mapping_test.bal:41]: Can't we use client data-binding here ? Check for other places as well.

Comment (TharmiganK) [reply ballerina-tests/tests/service_dispatching_header_param_binding_ideal_mapping_test.bal]: Resolved in [`dfed505`](https://github.com/ballerina-platform/module-ballerina-http/pull/653/commits/dfed50595eb141f7941d3405f50e6e299374c288)

**Review by chamil321 — APPROVED** [review]

**Pull request merged**